### ISSUE TITLE
Some CSG nodes (including CSGBox3D, CSGTorus3D, and CSGCylinder3D) have incorrect mesh UVs

### ISSUE THREAD

**Issue by Meorge** [godotengine/godot#95834]
### Tested versions

- Reproducible in v4.3.stable.official [77dcf97d8]

### System information

Godot v4.3.stable - macOS 14.6.1 - GLES3 (Compatibility) - AMD Radeon Pro 5500 XT OpenGL Engine - Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz (16 Threads)

### Issue description

CSGBox3D nodes have UV mapping so that textures can be displayed on them via their material, but it appears none of the faces have a completely correct mapping, so an asymmetrical texture is always rotated or reversed in some way.

https://github.com/user-attachments/assets/e1532440-0eff-46dd-a699-575bf2a9d5b0

If the box's X scale is set to -1, then the texture on the face with the normal pointing in the +Z direction does appear correctly. 

https://github.com/user-attachments/assets/212dc71b-3b03-4983-98f7-661abd9882ad

This seems like a strange change to have to make to the cube to get a correct mapping, though.

### Steps to reproduce

1. Create a 3D scene.
2. Add a child node of type CSGBox3D.
3. Create a StandardMaterial3D on the Material property of the CSGBox3D.
4. Add an asymmetrical texture to the Albedo Texture property of the material.

### Minimal reproduction project (MRP)

[bad3d.zip](https://github.com/user-attachments/files/16667341/bad3d.zip)

**Comment by The-Cyber-Captain** [comment]
Confirmed (also in 4.4-dev). I happened to have a scene set up for examining the default mesh-primitive UVs when I saw this, so dropped in two CSGs and nabbed the following screenshots in a bid to further demonstrate @Meorge 's issue.

See CSGBox (boxCSG,png) vs quadMesh, planeMesh, and - _arguably_ erroneous in a different way, which is what I'd set up to check - boxMesh. Anyway, echoing Meorge's issue; doesn't seem that there's any real logic to CSGBox's UVs being generated in the current orientation. It's certainly not intuitive to any prototyping end-user / game-developer.

Also... won't it risk a heap of confusion when any 'final' CSG is exported out for proper retopology - or modelling - in a third party package?

<img width="1269" alt="boxCSG" src="https://github.com/user-attachments/assets/70143e23-0687-47ee-8352-d2a04e0a6e14">
<img width="1269" alt="quadMesh" src="https://github.com/user-attachments/assets/3074ae88-3ff5-40b7-b923-23a5f4f23c05">
<img width="1269" alt="planeMesh" src="https://github.com/user-attachments/assets/8d8c7ce6-29af-4667-8fce-1f057e194444">
<img width="1269" alt="boxMesh" src="https://github.com/user-attachments/assets/48d393c0-86c8-4bc3-a8df-1e9dfec6d597">

**Comment by The-Cyber-Captain** [comment]
Notably (though, of course, what is 'correct' unwrapping becomes more esoteric, and perhaps less relevant - especially with intended CSG use beyond the box shape) this same 'backwards' UV approach appears to apply to CSGCylinder...

See: CSGCylinder, CSGCylinder (64) and cylinderMesh

<img width="1269" alt="cylinderCSG" src="https://github.com/user-attachments/assets/2fb04557-ee88-4d98-9045-586b2e7694d0">
<img width="1269" alt="cylinderCSG64" src="https://github.com/user-attachments/assets/88d52b93-d415-4125-9f2a-e6a43a2a7912">
<img width="1269" alt="cylinderMesh" src="https://github.com/user-attachments/assets/f7f641fd-cb7b-450a-b50c-8bc922ac2343">

**Comment by The-Cyber-Captain** [comment]
... however, CSGsphere maps exactly as the sphereMesh primitive! I wasn't going to include these (admittedly large!) images for that case initially. But, reconsidering, the _inconsistency_ suggests at the very least CSG UVs are revisited... with Box and Cylinder adjusted to rectify as per @Meorge 's initial logged issue.

CSGSphere, CSGSphere (64), sphereMesh.

<img width="1269" alt="sphereCSG" src="https://github.com/user-attachments/assets/ae416faf-de37-41f3-8f58-90a337df25a9">
<img width="1269" alt="sphereCSG64" src="https://github.com/user-attachments/assets/d337c219-8fe0-4a00-888e-794f557b30af">
<img width="1269" alt="sphereMesh" src="https://github.com/user-attachments/assets/36642a7d-ca01-4625-a76f-77ce76ca3f3c">

**Comment by The-Cyber-Captain** [comment]
For completeness....

* CSGMesh3D works as one would hope; brings over the UV from source mesh.
* CSGPolygon3D; adapts UV in a sensible fashion for points in XY plane on front face. 'Extrudes' 0 to 1 in depth.

....

**Comment by The-Cyber-Captain** [comment]
... but, erm, torus?

Yeah. OK. I *think* I see what you're doing torus. LOL.

CSGTorus, CSGTorus64, torusMesh

<img width="1269" alt="torusCSG" src="https://github.com/user-attachments/assets/03c661a0-461c-4ef5-86a3-47a09859b4f3">
<img width="1269" alt="torusCSG64" src="https://github.com/user-attachments/assets/3e84b7fd-d3c0-4a1e-b54e-4ef5f825d5d0">
<img width="1269" alt="torusMesh" src="https://github.com/user-attachments/assets/cba69d12-9365-46cc-929d-6791dea29c0b">

**Comment by The-Cyber-Captain** [comment]
TL;DR?

I suggest we update the issue name to - at least - include Torus (which is simply wrong by any metric), as well as Cylinder (which exhibits same problem as Box). If not, that is, simply to 'Refactor of broken CSG UV maps' - which is hopefully now fully defined above. //crosses fingers

**Comment by Meorge** [comment]
Thanks for finding those! I'll make that update to the issue name 😄

**Comment by Calinou** [comment]
Thanks for the report! Consolidating in https://github.com/godotengine/godot/issues/19143.
